app/helpers/react_on_rails_helper.rb

```diff
@@ -2,132 +2,196 @@
+  #      a generator:
+  #    For client rendering, expose the react_component_name on window:
+  #      window.MyReactComponentApp = MyReactComponentApp;
+  #    For server rendering, export the react_component_name on global:
+  #      global.MyReactComponentApp = MyReactComponentApp;
+  #    See spec/dummy/client/app/startup/serverGlobals.jsx and
+  #      spec/dummy/client/app/startup/ClientApp.jsx for examples of this
   # props: Ruby Hash which contains the properties to pass to the react object
   #
-  # Naming Conventions:
-  # Suppose your component is named "App" (can be anything)
-  # 1. Inside app/startup/ServerApp.jsx, setup the component for server rendering. This is used by
-  #    the global exports in step 2
-  # 2. Component_name is changed to CamelizedUpper for the exposed component,
-  #    so we have App as the exposed global (see app/startup/serverGlobals.jsx)
-  # 3. Inside app/startup/ClientApp.jsx, you want to export a function that generates
-  #    your component, and when you generate the component, you must use this naming convention
-  #    to get the data.
-  #    App(__appData__)
-  #    That way it gets the data you pass from the Rails helper.
-  #
   #  options:
-  #    prerender: <true/false> defaults to true
-  #    trace: <true/false> defaults to false, set to true to print additional info
+  #    prerender: <true/false> (configurable default), set to false when debugging!
+  #    trace: <true/false>  (configurable default), set to true to print additional info
   def react_component(component_name, props = {}, options = {})
+    # Create the JavaScript and HTML to allow either client or server rendering of the
+    # react_component.
+    #
     # Create the JavaScript setup of the global to initialize the client rendering
     # (re-hydrate the data). This enables react rendered on the client to see that the
     # server has already rendered the HTML.
-    @react_component_index ||= 0
-    prerender = options.fetch(:prerender) { ReactOnRails.configuration.prerender }
-    trace = options.fetch(:trace, false)
-
-    dataVariable = "__#{component_name.camelize(:lower)}Data#{@react_component_index}__"
-    reactComponent = component_name.camelize
-    domId = "#{component_name}-react-component-#{@react_component_index}"
-    @react_component_index += 1
-
+    # We use this react_component_index in case we have the same component multiple times on the page.
+    react_component_index = next_react_component_index
+    react_component_name = component_name.camelize # Not sure if we should be doing this (JG)
+    dom_id = "#{component_name}-react-component-#{react_component_index}"
+
+    # Setup the page_loaded_js, which is the same regardless of prerendering or not!
+    # The reason is that React is smart about not doing extra work if the server rendering did its job.
+    data_variable_name = "__#{component_name.camelize(:lower)}Data#{react_component_index}__"
```

Comment:
@mapreal19 Just a styling thing so that the name always looks consistent. I think originally, the coder had to know of the the name, but that's no longer true. Still, for `MyBigComponent`, the name becomes `__myBigComponentData__` for the data variable. If there's a good reason not to change the name, we could do that.